### PR TITLE
Add AAD support for EG

### PR DESCRIPTION
**Pull request by rakshith91**
fixes https://github.com/Azure/azure-sdk-for-python/issues/17963

### CODE REVIEW

File: sdk/eventgrid/azure-eventgrid/README.md [hunk 1]
```diff
@@ -38,6 +38,34 @@ az eventgrid domain --create --location <location> --resource-group <resource-gr
...
 
+#### Using Azure Active Directory (AAD)
+
+Azure Event Grid provides integration with Azure Active Directory (Azure AD) for identity-based authentication of requests. With Azure AD, you can use role-based access control (RBAC) to grant access to your Azure Event Grid resources to users, groups, or applications.
+
+To send events to a topic or domain with a `TokenCredential`, the authenticated identity should have the "EventGrid Data Sender" role assigned.
+
+With the `azure-identity` package, you can seamlessly authorize requests in both development and production environments. To learn more about Azure Active Directory, see the [`azure-identity` README](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md).
+
+For example, use can use `DefaultAzureCredential` to construct a client which will authenticate using Azure Active Directory:
```
Comment (swathipil) [inline sdk/eventgrid/azure-eventgrid/README.md:49]: typo, nit:
```suggestion
For example, you can use `DefaultAzureCredential` to construct a client which will authenticate using Azure Active Directory:
```

**Review by swathipil — COMMENTED** [review]
small nit, but o/w lgtm

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_helpers_async.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+from azure.core.pipeline.policies import AzureKeyCredentialPolicy, AsyncBearerTokenCredentialPolicy
+from azure.core.credentials import AzureKeyCredential, AzureSasCredential
+
+from .. import _constants as constants
+from .._signature_credential_policy import EventGridSasCredentialPolicy
+
+def _get_authentication_policy_async(credential):
```
Comment (yunhaoling) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_helpers_async.py:11]: nit: I think we could avoid duplicating the `_get_authentication_policy` method for async by adding a credential type argument in the sync helper.

```python
def _get_authentication_policy(credential, bearer_token_policy=BearerTokenCredentialPolicy):
    if hasattr(credential, "get_token"):
        return bearer_token_policy(
            credential,
            constants.DEFAULT_EVENTGRID_SCOPE
        )
```

then in the async, we do
```python
_get_authentication_policy(credential, AsyncBearerTokenCredentialPolicy)
```

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_helpers_async.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    if credential is None:
+        raise ValueError("Parameter 'self._credential' must not be None.")
+    if hasattr(credential, "get_token"):
```
Comment (yunhaoling) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_helpers_async.py:14]: dumb question: what error would be raised if a user passed a sync AD credential into the async client?

File: sdk/eventgrid/azure-eventgrid/README.md [hunk 4]
```diff
@@ -38,6 +38,34 @@ az eventgrid domain --create --location <location> --resource-group <resource-gr
 In order to interact with the Event Grid service, you will need to create an instance of a client.
 An **endpoint** and **credential** are necessary to instantiate the client object.
 
+#### Using Azure Active Directory (AAD)
```
Comment (JoshLove-msft) [inline sdk/eventgrid/azure-eventgrid/README.md:41]: Nice, this reminds me I need to add this for .NET 😄

File: sdk/eventgrid/azure-eventgrid/README.md [hunk 5]
```diff
@@ -38,6 +38,34 @@ az eventgrid domain --create --location <location> --resource-group <resource-gr
...
+
+Azure Event Grid provides integration with Azure Active Directory (Azure AD) for identity-based authentication of requests. With Azure AD, you can use role-based access control (RBAC) to grant access to your Azure Event Grid resources to users, groups, or applications.
+
+To send events to a topic or domain with a `TokenCredential`, the authenticated identity should have the "EventGrid Data Sender" role assigned.
+
+With the `azure-identity` package, you can seamlessly authorize requests in both development and production environments. To learn more about Azure Active Directory, see the [`azure-identity` README](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/identity/azure-identity/README.md).
```
Comment (johanste) [inline sdk/eventgrid/azure-eventgrid/README.md:47]: We shouldn't link to the old `main` branch.

Comment (rakshith91) [inline sdk/eventgrid/azure-eventgrid/README.md:47]:
```suggestion
With the `azure-identity` package, you can seamlessly authorize requests in both development and production environments. To learn more about Azure Active Directory, see the [`azure-identity` README](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/identity/azure-identity/README.md).
```

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/README.md]: good catch - updated

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 6]
```diff
@@ -22,20 +22,24 @@
...
 from .._policies import CloudEventDistributedTracingPolicy
 from .._models import EventGridEvent
 from .._helpers import (
-    _get_authentication_policy,
     _is_cloud_event,
     _is_eventgrid_event,
     _eventgrid_data_typecheck,
     _build_request,
     _cloud_event_to_generated,
+    _get_authentication_policy
```
Comment (johanste) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:35]: Nitpick: missing trailing comma - did you run this through black?

Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:35]: nope - i did not - can do that

File: sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py [hunk 7]
```diff
@@ -73,7 +78,7 @@ class EventGridPublisherClient:
     def __init__(
         self,
         endpoint: str,
-        credential: Union[AzureKeyCredential, AzureSasCredential],
+        credential: Union["AsyncTokenCredential", AzureKeyCredential, AzureSasCredential],
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/azure/eventgrid/aio/_publisher_client_async.py:81]: That seems inconsistent, if not used those types should be in the `TYPE_CHECKING` as well, but I see now reason why some type would be string and some would be types

File: sdk/eventgrid/azure-eventgrid/tests/asynctestcase.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    def __init__(self):
+        self.token = AccessToken("YOU SHALL NOT PASS", 0)
+
+    async def get_token(self, *args):
+        return self.token
+
+
+class AsyncEventGridTest(EventGridTest):
+
+    def generate_oauth_token(self):
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/tests/asynctestcase.py:29]: You shouldn't need that, there is everything you need in devtools to care care of that for free

File: sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py [hunk 9]
```diff
@@ -94,4 +105,28 @@ def _get_resource_group(self, **kwargs):
...
+class EventGridTest(AzureMgmtTestCase):
+    FILTER_HEADERS = ReplayableTest.FILTER_HEADERS + ['aeg-sas-key', 'aeg-sas-token']
+
+    def __init__(self, method_name):
+        super(EventGridTest, self).__init__(method_name)
+
+    def get_oauth_endpoint(self):
+        return os.getenv("EG_TOPIC_HOSTNAME")
+
+    def generate_oauth_token(self):
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/tests/eventgrid_preparer.py:121]: Same, you don't need that

File: sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py [hunk 10]
```diff
@@ -343,5 +344,19 @@ def test_send_custom_schema_event_as_list(self, resource_group, eventgrid_topic,
 
     def test_send_throws_with_bad_credential(self):
         bad_credential = "I am a bad credential"
-        with pytest.raises(ValueError, match="The provided credential should be an instance of AzureSasCredential or AzureKeyCredential"):
+        with pytest.raises(ValueError, match="The provided credential should be an instance of a TokenCredential, AzureSasCredential or AzureKeyCredential"):
             client = EventGridPublisherClient("eventgrid_endpoint", bad_credential)
+
+    @pytest.mark.live_test_only
```
Comment (lmazuel) [inline sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py:350]: why live only?

**Review by swathipil — APPROVED** [review]

File: sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py (hunk 10 above)
Comment (rakshith91) [reply sdk/eventgrid/azure-eventgrid/tests/test_eg_publisher_client.py:349]: little tricky to generate recording given we use resource group preparers and envvars for secrets

**Pull request merged**